### DISCUSSION TITLE
Outline View References or Peek References

### DISCUSSION THREAD

**Discussion by thomprl** [codefori/vscode-ibmi#1533] [General]
One thing that really holds me back on using VSCode is some of the limitations on the outline view. When looking at large programs it’s very helpful to see the lines a variable is used on and which lines are setting the variable to a new value. This allows you to jump to that code quickly and know where to set breakpoints. This may not be possible in VSCode. But I’m RDi where each of the lines that have an M beside them in the screen shot the variable is being changed. I find the peek a little cumbersome to use. I know VSCode can’t be RDi though and maybe I’m missing something because I’m not a daily user of VSCode. 

![image](https://github.com/halcyon-tech/vscode-ibmi/assets/17936568/0151bbce-9d10-46e1-b91d-3f013ea33e09)

**Comment by Gregory1701** [comment]
@sebjulliand @worksofliam - Hey guys, I've often thought this same thing, but never got around to starting a discussion. Can this be done? Thanks @thomprl for starting the conversation!

**Comment by thomprl** [comment]
Just wondering if this is on the radar for anytime in the future?

**Comment by mlitters** [comment]
Hi there,

I finaly got the Outline to work after deactivating the IceBreak RPG Add On in VS Code and activate the RPG Language Add On. Maybe this is an important hint for other that have the same problem.

BUT

is this really everything we can use in Code4i as the Outline?

![image](https://github.com/codefori/vscode-ibmi/assets/24865681/3cbecf2f-6500-4dfb-bc45-74b016f26191)

As @thomprl already mentioned is that the outline is a very important feature at least in existing, monolithic RPG programs. Doesn't matter if column based, full free or total free. thousands of lines of (unknown) code is not easy to understand and the outline feature in RDi is for me the best benefit of that tool.

So is there a plan to enhance it or can we enhance it somehow? I have multiple Developers in my Team who should be able to do it but I don't want to spend time into it, if there is already someone else working on it. 
So a quick feed back would be very helpful.

Many thanks for the great work on the multiple add ons and go on!

Kind regards
Markus

**Comment by thomprl** [comment]
I wanted to bump this again and see if anything similar could be added?

**Reply by worksofliam** [reply]
Going to tag @mlitters so he can see this too.

VS Code will never show references in the built in outline view. Every language in VS Code builds to a common API and therefore the built in outline view is the same for every language - none of them show references.

Of course, you are still able to find references to variables by using _show all references_, _go to references_ or _peek references_.

It's not like RDi and I'm okay with that. If we want to rebuild the RDi outline view, then we'll have to create our own view. It's doable, but not very high priority for us since we are more interested in fitting in with VS Code.

I need to chat with some other folks to see if this RDi-style outline is worth our time.

**Reply by thomprl** [reply]
It’s probably the biggest hurdle with switching from RDi to VSCode. Older programs were just built different and not modularized so you have 1000s of variables and needing to quickly navigate a program RDi outline view is killer for that.  If you know a variable is getting set to a wrong value you can quickly find all the places it’s being set. 

I do know VSCode is not RDi so I understand.

**Reply by worksofliam** [reply]
@thomprl VS Code has similar language features to RDi, they just work differently.

https://code.visualstudio.com/docs/editor/editingevolved#_peek

**Reply by mlitters** [reply]
I total agree with Liam. When you create an actual RPG program with many procedures and NO Display Files and NO old Physical Files with native I/O and NO Printer Files you really don't need such an Outline we are just creating.
In this case the VS Code functions are much more helpful because the (most times younger) developers, coming from Javascript, PHP, c#, Rust or whatever language, using VS Code, are used to use it in this way and so they (and we and ME!!!!) should also be able to use it.

BUT

the other side of the coin is, that there are many, many, many, many... ups - endless loop with goto?!? - old RPG Programs and a lot of developers are faced with the fact to maintain them.
Even read that code and understand it is very, very hard and can take up many days for only one program.
When you use old DSPF, PRTF, native I/O, you can have hundreds and thousands of fields - all global. 
Therefore the classical Outline is very helpful.
We added already some more cool features like opening the source code of any procedure you are using within your code as well as the CALLed programs, or DSPF, PRTF, etc..
We will add a lot of more features you will need at your daily work when you are working with such old code.
So for me it is not a question about Black or White - you can and should use both options if you need them.
When you create new things, PLEASE, use the actual possibilities in RPG. It is still a great, robust, secure, scalable, fast and powerful language and in many cases it still makes sense to use it.

**Comment by ewarwoowar-git** [comment]
I've just watched the recording of Friday's Code for i Fridays and was impressed, if not encouraged by the demo of the Outline view.

An outline view which looks and works like the one in RDi (or the lack thereof) if one of the blockers for our devs to make the jump from RDi, so it's nice to know that something is in the pipeline.

**Reply by mlitters** [reply]
Many thanks for the feedback. We got more and more of this which motivates us to create a V1 even faster.
Beside the RDi features we add many other things, IBM is not willing to add to it's outline. Some of them you can see it in the Video. Others will follow.

**Comment by thomprl** [comment]
Yeah I just watched it as well, this is exactly what is needed.  The Peek/References just isn't good enough when you are used to using RDi.

**Reply by chrjorgensen** [reply]
>The Peek/References just isn't good enough **_when you are used to using RDi_**.

Hm, isn't this one of the situations where you have to get out of your comfort zone to gain some advantages?

As said in several comments, VS Code is not RDi - and will never be. It's a completely different tool with new ways of doing stuff. The Peek feature is a another way of viewing definitions and references, totally different from the Outline view in RDi. The more I use peek in VS Code, the more I like it, because peek allows you to quickly jump around in the sources without leaving your current position. And VS Code can generate a search list for the references to an item, if you want a more permanent view of the references (Find All References). 😍 

Please correct me if I'm wrong, but I sense that the request for an RDi-like Outline comes from developers maintaining old, monolithic RPG/400/III/II programs with hundreds or thousands of global variables. I'm happy to not have to maintain such programs anymore, but pushing for having built-in features for maintaining old source in a new, modern editor developed for modern development style is not the right way IMHO. Such specific requirements are better done in extensions (separate from the editor) like the one demo'ed by Markus Litters in the latest Code for i Friday meeting. 🤔 

Switching to a modern editor could be seen as an opportunity to also switch to modern code styles - and maybe even modernize the old code or slowly pick it apart to avoid having this heavy maintenance work in the future. If you keep the old code as is, you could just as well keep the old editor. Switching editors always mean changing habits and climbing a learning curve, but mostly it is worth the effort once the FUD is overcome. Things may be different, but if the new features are worth it it's a win. If not, the old editor is still there. 🎉

**Reply by mlitters** [reply]
Hi Christian,
first of all:
thanks god  VS Code is not RDi!!!!!

I can't tell you how many hours in my life I have lost because of RDi problems and my team will sign that too as well as most of our customers.
So I see VS Code and Code4i as an alternative for RDi and not "only" for modern RPG Development.

I am doing RPG Microservices with Niels Liisbergs VS Code Add On for IceBreak for more than 7 years now and I can only push every RPG developer to look on this cool way of programming - now also with Code4i.
My Team which is growing very fast, has many (young and very young) c#, Javascript, PHP and other Developers and they can read and write those RPG "Microservice" programs very fast and with VS Code they are all familiar.

But there is also the other side of the coin.
Most of our customers still have thousands of old RPG Programs and it is not easy (in many cases not possible because of time and money) to convert that to something modern.
If you have a column based spaghetti Code with 2000 global variables and convert it to total free (if it is really possible because of all the goto's....) then you have a total free spaghetti Code with 2000 global variables.
And many times it take too much time to modernize everything or at least parts so you keep it running because there are other tasks to do... 
Therefore we develop our AI based RPG Analyzing tool which analyzes such old RPG code and explain it to us and everybody and then you can also convert it to something better (modern RPG, Javascript, Rust, c#, or whatever).

One step toward this was creating a component which get all the information out of the old code like which fields are used in which file (DSPF, PRTF, PF, LF), which Datastructures and where and so on - then we saw the result looks like the existing Outline in RDi and because we need to work a lot with such old code you don't need to maintain (you should thank god for that every day!!!), this new Outline Add On will help us a lot as well as hundreds of other RPG Developers so we decided to make it a VS Code Add On.

I think it is important to face the reality and that is, that out there are many companies still running RPG-II and OCL Code. There are many running RPG-III / RPG/400 code and there are many running column based ILE Code that don't deserve the name "ILE" because it is just converted old code with all it's problems.
BUT - all of these programs normally have one thing in common: They work and keep the companies up and running.
We shouldn't forget that.

Our mission is to help all the developers as well as these companies who collected a lot of technical debt over decades out of this dead end road.
From this perspective I think it is very good and important that the users can choose between different Add Ons for VS Code (as well as for RDi - think at the PDM perspective...).
We are very happy that Code4i is now able to handle column based code in a better way because the developers need an alternative to RDi.

The young ones I train all say that they can arrange with RPG but not with RDi.... that is a fact.
So go on and make VS Code the better RPG IDE than all others.

By the way other RPG IDE's: Our new Outline Add On has the Project Codename "Morgana"
Some think because of my Spanish Galgo which has the same name....
But the truth is more complex ;-)

**Reply by chrjorgensen** [reply]
Hi Markus.

Great to see you here - I still remember us chatting at CEC2024 in Milan! 😃 

We do not disagree at all...! 😃 
I know not everyone is as fortunate as I am... they still have to maintain monolithic RPG programs and need some special tools to handle these kind of sources. You probably see more than your share of such code - but at least you have made it a living! 😉 

My arguments were sparked by the quote in my first post - _that the peek references in VS Code isn't good enough when you're used to the outline in RDi_. I've used RDi for many years, ever since it was called WDSc (WebSphere Development Studio code) and I'm happy to report that it is much better now! 😅 I've also used VS Code for some years, which is way better than RDi - and in my experience the RDi outline was not any better than the VS Code outline, on the contrary. The RDi outline feels cluttered when I compare the two, because of all the line numbers next to each variable. VS Code outline is more dense and does not show any line numbers, just the location (e.g. procedure, file) of the variable definition.

To see all the code where an item is referenced, you don't use the outline but instead the peek or find references feature in VS Code. This is quite handy when you get used to it, since you see the source lines with the item, and a click will show the lines before and after the line with the reference. You can quickly spot how the item is used - if its before or after an equal sign, or if it's in the Factor1, Factor 2 or Result column for fixed format RPG. And the peek feature let you stay where you are in the code while showing all this, you don't have to navigate back to your original location. Quite nifty IMHO and I find myself only using the outline for navigating to procedures in the code. 😃 

I know that this discussion is about the outline view and monolithic sources - I just challenge the claim that the VS Code outline (and peek references) is not good enough for this situation and blocking developers from switching to VS Code from RDi. Every tool has its strengths and weaknesses, and even if the VS Code outline would be considered weaker than the one in RDi, VS Code would still has the advantage over RDi - if not we would not see the many users coming to VS Code from RDi.

If a developer has tried VS Code for some time and then decides that there is no productivity gain because of the outline differences, then he/she has some options:

#### 1. Switch back to RDi
This would be the easy way out - nothing new to learn and everything is like it used to be. And RDi is improving - albeit slowly and nowhere near in the speed of VS Code and Code for IBM i.

#### 2. Ask the Code for IBM i developers to create the feature
This is an option but does not guarantee the feature will ever be created. Code for IBM i is an open source project maintained by volunteers in their spare time - it is not an IBM product, as seen by many in the IBM i community. Liam is the only developer being employed and paid by IBM (completely fair, since this is his creation!), but the rest of us are developing, testing and supporting in our spare time and not getting paid for this. We will prioritize what we see as benefiting the majority of our users - and what is most fun! 😄 

#### 3. Clone the source for Code for IBM i and develop the feature yourself
This is an option but not used very much, at least to my knowledge. Anyone can fork the project and make their own version, but this requires quite some TypeScript programming experience, not usually found in the IBM i community.

#### 4. Create the feature as an add-on to Code for IBM i
**This is where you come in, Markus!** I applaud your efforts to create an extension for Code for IBM i, providing all the features you feel are missing in the VS Code outline and in the RDi outline. This is a cool project, and I watched your demo last Friday and am eager to see what it can do and how you did it... 😍 When the source is published, I will probably be one of the first outsiders to examine your code... ❤️ 


Keep up the great work you're doing - for VS Code and for the IBM i community!
I'm looking forward to see you next time, whenever that will be. 😃

**Reply by ewarwoowar-git** [reply]
Hi Christian,

Your assertion that VS Code is way better than RDi is an opinion, and is one I would like to share, but unfortunately, as a long-time user of RDi (I've used it since 2006 when it was called WDSc - and _that_ is a whole other story!) I don't feel that VS Code is quite 'there' yet...but it's definitely improving and I want me and my team to make the leap as soon as we can as I can certainly see the benefits.

Your arguments run parallel to my own with regards to trying to convince my team to make the leap from SEU to RDi - we have a couple of devs who still bemoan having to use RDi and go on and on about how coding in SEU is 'better'! - I guess it's a case of 'when you know, you know', so I'm aware that, to you, I must appear to be like my SEU-aficionados.  I find it interesting that those devs have not had much experience with other languages or the modularity of modern ILE and struggle to understand some of the more modern enhancements to RPG or the necessity to use them.

We also have a couple of devs who have used VS Code for other languages who also believe that VS Code is better than RDi, however they still revert to using RDi to maintain larger programs (because of the Outline view - hence my reply) and coding green screens.

Personally, I've used a number of languages over the years, and have used Visual Studio for a number of years doing C# development, so like the similarities with VS Code and really want to make the move to it, but because we support (and continue to develop) an old application with many thousands of programs and green screens (most are monolithic monsters!) the lack of an RDi-like Outline view really is a blocker for us (as is a lack of a screen designer).

Markus' Outline View extension is great news for us as it will definitely remove one barrier (perceived or otherwise) - now we just have to hope some talented people take it upon themselves to code a green screen designer).

@mlitters - I echo Christian's sentiments and thank you for what you are doing.

**Reply by mlitters** [reply]
Wow - now we have a real cool discussion about so important things.
That is very good and I receive so many emails in parallel about this topic. We already had a very powerful feedback in Germany and now also in the US.
Thanks to everybody who is responding and thinking about this topic.

@chrjorgensen 
I am not sure if I already mentioned that we are working really a lot with VS Code in many different languages and environments (there are so many wonderful add ons right now) so we really know and like the way VS Code is "normally" doing things like finding references and implementations.

Regarding to RPG we are talking about two different RPG's. 
The old one and the newer one.
With the newer one, which I created since the first day IBM announced ILE RPG back in the middle of the 90's, we can handle everything you mention. As already mentioned we are doing a lot of things with noxDB and so called Microservice RPG which is really cool because we can use all of our procedures from everywhere - in all kind of applications that support http requests.

I am on the platform since 1990 and started with RPG-II in /36 Mode and was really shocked in these days because in that time I was working on my very cool Commodore Amiga 2000 with real Multitasking, powerful graphics, stereo sound and programming in c++ and a real cool Basic Compiler version that was very easy and fast.
Then they put me in front of a Twinax TErminal with 5250 Screens and that - what is that - code. Looks like Assembler I said.
Funny thing is: I teach a lot of people in RPG and there are many younger people of course and most of them have the same reaction but more than 30 years later.... 
If one told me that I will working with RPG-II until the rest of my life I think would have said: NO thank you. I go and be a Gardener.

Fact is: Out there are still (34 years later.... )a lot of old, very old and very, very old RPG applications and for those we are doing this stuff because more and more RPG developers retire and I am one of the youngest people on earth that still know RPG-II.
I teach my 23 years old Team members to understand it as well and Gerhard and his Team is working on our Hybrid AI Tool that can also understand that (as well as RPG-III and ILE RPG) "Spaghetti Goto Mess" - that's how I call it.

As you can see on the reactions here and I can see on the reactions I get on other media there are many people who really need this and many of them have no choice to convert/modernize the existing code base. We are talking about several thousand of old RPG programs in these companies.
The oldest application we have at our customers that is still running daily is now 50 years old. It comes from the /32 system and still runs with only a few changes in all these years.
The former Owner of the company, who "programed" that himself, always said: It works so why should I spend money to change it.... 
That is the curse of the IBM backwards compatibility.
The blessing of this is that so many companies are still on IBM i.
I guess a lot more would have left the platform if IBM really forced them to modernize a lot in the past.

But it is as it is and now we are all sitting in this boat.
My team and me are looking forward to find solutions for this challenge and it looks like we are on a very good way.
The Outline is "only" a side effect of our main toolset we are developing. With this we already have very good results in analyzing old (and very, very old) RPG code and it also gives us much more transparency with all this Goto Sh.... and it's 99 indicators who are in front of many Goto's.... 

So go on with the modern stuff. We need it as well and we use it and tell everybody to use it - if it is possible.

@ewarwoowar-git 
Thanks for your words as well. 
It looks like we have very similar experiences while you are even longer in this "circus" than me ;-)

You mentioned something that we face also many times. There are still a lot of RPG developers using PDM and SEU.
Also people who are in the mid 50's.
We have a Workshop called "Evolution of RPG". In this I explain the people how RPG evolved over the decades and there are still many people that are really positive shocked that NOW (!!!) it is possible to use real date fields in RPG and you can calculate with them with the %BIF's....
I tell them that this feature is of course "brand new" because it exists since 1995 (as far as I remember...) so when should they heard about that :-D
What else should I say?!?

That is the reality we are facing many times.
Thanks god we also have customers that created rules and naming conventions back in the 70's/80's (yes - it was possible also with RPG-II and RPG-III!!!) and all of the developers followed these. With such applications it is no problem to transform or modernize them.
These people normally did that already.
One also created a Database Access Layer in RPG-III and that works very well!

But we also have customers with 1 or 2 developers that need to do the work for 4 or 6.... 
And I can understand that after many years of arguing against walls that they only do their daily work and then go home and try to forget all of this....
Not everybody is a born developer. For many people it is a job and they don't want to spend their free time for learning.
I do that as well as many of you of course but not everybody.
Please keep that in mind.

I will be happy to read more opinions and experiences because this is important for a community to not forget who is working within it.

**Reply by chrjorgensen** [reply]
Hi Paul.

Seems like we've been on the same path through the years - I started coding RPG II in 1986 and started using WDSc as well... 😄 

It's funny: I actually thought about arguing that SEU is way better than RDi at CL prompting and it didn't stop me and others switching to RDi back then - and now you're telling us about your SEU aficionados! 😆 

Well, technology evolves and IT technology evolves very fast and we all have a hard time keeping up. I'm a curious person and like exploring the new possibilities in the new technologies and thus have switched tools and languages a few times during my career. I will never understand fellow developers who do not want to learn new technologies - they put themselves in a corner eventually, with no way out.

Code for IBM i (just called "C4i" within the core team) was developed for modern and fast RPG development, and as a core team member it is my opinion(!) that it should stay like that and not be cluttered with code, options and views to support old technology, which should have been modernized years ago. This is probably not a popular opinion amongst developers maintaining "monolithic monsters" and green screens... but the opposite opinion is also not popular, just in another - I think and hope larger -  segment of developers, who do modern RPG development.

The resources within the project is limited, and it will always be a matter of prioritizing where to put the resources - developing features for the modern RPG code or features for the older RPG code. We have had quite a few of these issues: SEU colour codes support, line numbers, source line dates etc. How much effort should the C4i team spend to support old style RPG programming using old style source code (members) versus new style RPG programming and new style source code (IFS streamfiles, local files, git)? My vote goes to the new style RPG programming, because that is the new norm. (I only use the term "new" in the context of RPG history - it's not new anymore, free-format RPG and procedures have been available for a decade and more!)

This is why I really like what @mlitters and his team are doing - adding the support on top of C4i! This expands the resources for all of us, making C4i supporting the new code and the add-on support the old code.

Well, I better say it - the above is just my opinion... 😆 

I wish you all a great day, a merry Christmas and a Happy New Year! ❤️

**Reply by ewarwoowar-git** [reply]
Hi Chris,

Yes, it sounds like we have been on a parallel journey!  

And to be honest, I'm very much aligned with your thoughts and aims.  The only thing is that my reality is considerably different to yours: I am in charge of supporting a system which is over 25 years old, and had been presided over by a small team of developers who have largely been hiding under a rock...until I came along and kicked that rock! 🤣 

It's been an uphill battle to introduce ILE concepts such as modularity and service programs, and modern design concepts such as having a data access layer using service programs and embedded SQL, but they're getting there.  And don't event get me started on flat file interfaces!

As I said, I have struggled to convince some devs that using SEU was holding them back from becoming better developers (not to mention their code from becoming easier to maintain), and finally convinced the last luddite to move to RDi earlier this year (after coaxing him for nearly 3 years without success), so from my viewpoint, in order to move on from RDi to VS Code (which I do want to do), it would be easier to lead them from something they know towards something they don't, and providing a familiar interface/process/whatever would  be a really great help in achieving that.

I just wish I had the skills to be able to contribute (still don't know how to use git yet!) - maybe in the future when I can find some time!

**Reply by chrjorgensen** [reply]
@mlitters You just beat me in the response time competition! 😆 

I fully agree this is an important discussion - we all learn by this. getting viewpoints from outside our own spheres. 👍 

Code for IBM i will never be a perfect tool and will never include all features to support all kinds of RPG. As a VS Code extension it is based on the idea that additional features should be brought by additional extensions, splitting the code and functionality into several smaller components easier to maintain. It should not go in the footsteps of RDi and be a swiss-army-knife with all bells and whistles for IBM i development - this would make it difficult to maintain and slow down the evolvement, just like RDi.

It is not important who develops the extensions for new features. C4i is free and open source and exposes API's to allow easy creation of add-ons. Just don't expect the C4i team to bring all features... 😃

**Reply by mlitters** [reply]
@chrjorgensen You beat me in being member in the RPG club - so I guess we are fine 😉

I guess you brought it to the point.
The developers should be aware that there is a "core" C4i which is developed by some people and there are additional add ons that are developed by others and maybe also the same.
As far as I can see it, you are not angry when we find something in the core and fix it or make it better or give you some ideas and so on.
That is what I do for many years with IBM / Fortra to boost the RDi but you know the Wishlist is very long there and the list with bugs is from my perspective even longer....
Last week I had a training for RPG beginners and showed them RDi and told them about some of the issues they may have and in the middle of the training one of the participants had a new error. It showed the wrong Library list... never had that before and this time it was on Windows!!! 🤣

For us it is also not important who is doing the great stuff. It is important that it is done or in work.
My Team is very motivated with these topics and we go on and hurry. There is a big need for many things and IBM is not really boosting these things. So it is up to all of us.

@ewarwoowar-git 
If you want, we can show you how to use git with RPG. There are many ways in the meantime and even for your PDM/SEU colleagues Richard Schoen created a version that can handle green screen and classical source files (never needed that until now but good to know that there is something).
One of our next things we will publish will go into this direction.... but we are still very busy with it.

Thanks for the wishes. I wish you all the same. Freedom and health are the most important things!

**Comment by thomprl** [comment]
Didn't mean to get everyone riled up.  Hopefully Markus's Teams RPG Outline view will be the answer though.  Thanks to everyone that works on this project.

**Reply by mlitters** [reply]
Thanks for that and I hope that nobody will rile up!?!?!
For me we are all in the same boat.
There are a lot of old applications and a lot of RPG Developers in all ages, that need support and help to handle that old code because many times it is not that easy or even possible to modernize it - believe me - I am working in Modernizing for more than 30 years now - but they need to keep going and be transformed (someday).

Others are more lucky, like Christian and don't need to care about those old "Spaghetti Goto Mess Something" and I need to say that we also have customers that did it the right way and kept their code nearly up to date over the years.
There we are doing what Christian and Liam mentioned above.
Because we use VS Code a lot for Python, Javascript (Node.js, React, Angular whatever), Flutter, Rust and even c# we know how it works and all of us like that way.
For modern RPG Code we also want to use that.

But as mentioned there are many, many applications running since many decades and they need to run because they drive many businesses. 
That is something many sales people often forget.
I saw a lot of colleagues coming to some of my (later) customers telling them how easy they can modernize everything and crashed.
I saw a lot of companies trying to move away from the individual software they created on this POWERful platform and most of those projects crashed as well or got so expensive that they regret that step SOOO much.
I saw also companies that fail because of this. One of my Team members is from such a company so I can be thankful that they did it?!?

I hope nobody is angry because of our Outline Feature - it should help a lot of people and not be in competition with something else on the VS Code platform.
We want to make VS Code the bettest RPG IDE so the new (and most times younger) people feel more at home on IBM i.
RDi (I know it from the first Version as Christian and others mentioned - WDSC) is a big problem for the younger people and we need much more younger RPG developers when we want to go on with IBM i.
